opm/autodiff/ThreadHandle.hpp

```diff
@@ -0,0 +1,142 @@
+#ifndef OPM_THREADHANDLE_HPP
+#define OPM_THREADHANDLE_HPP
+
+#include <cassert>
+#include <dune/common/exceptions.hh>
+
+#include <thread>
+#include <mutex>
+#include <queue>
+
+namespace Opm
+{
+
+  class ThreadHandle
+  {
+  public:
+    class ObjectIF
+    {
+    protected:
+      ObjectIF() {}
+    public:
+      virtual ~ObjectIF() {}
+      virtual void run() = 0;
+      virtual bool isEndMarker () const { return false; }
+    };
+
+  protected:
+    class EndObject : public ObjectIF
+    {
+    public:
+      void run () { }
+      bool isEndMarker () const { return true; }
+    };
+
+    ////////////////////////////////////////////
+    // class ThreadHandleObject
+    ////////////////////////////////////////////
+    class ThreadHandleObject
+    {
+      std::queue< ObjectIF* > objPtr_;
```

Comment:
This will not be exception-safe I think, it should be possible to use unique_ptrs instead. It looks like you can move out of the front() of a queue and move into the queue with push().